src/coroutine/cancel.rs

```diff
@@ -0,0 +1,74 @@
+use crate::{ffi, Py, PyAny, PyObject};
+use futures_util::future::poll_fn;
+use futures_util::task::AtomicWaker;
+use std::ptr;
+use std::ptr::NonNull;
+use std::sync::atomic::{AtomicPtr, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+#[derive(Debug, Default)]
+struct Inner {
+    exception: AtomicPtr<ffi::PyObject>,
+    waker: AtomicWaker,
+}
+
+/// Helper used to wait and retrieve exception thrown in [`Coroutine`](super::Coroutine).
+///
+/// Only the last exception thrown can be retrieved.
+#[derive(Debug, Default)]
+pub struct CancelHandle(Arc<Inner>);
+
+impl CancelHandle {
+    /// Create a new `CoroutineCancel`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Returns whether the associated coroutine has been cancelled.
+    pub fn is_cancelled(&self) -> bool {
+        !self.0.exception.load(Ordering::Relaxed).is_null()
+    }
+
+    /// Poll to retrieve the exception thrown in the associated coroutine.
+    pub fn poll_cancelled(&mut self, cx: &mut Context<'_>) -> Poll<PyObject> {
+        // SAFETY: only valid owned pointer are set in `ThrowCallback::throw`
+        let take = || unsafe {
+            // pointer cannot be null because it is checked the line before,
+            // and the swap is protected by `&mut self`
+            Py::from_non_null(
+                NonNull::new(self.0.exception.swap(ptr::null_mut(), Ordering::Relaxed)).unwrap(),
+            )
+        };
+        if self.is_cancelled() {
+            return Poll::Ready(take());
+        }
+        self.0.waker.register(cx.waker());
+        if self.is_cancelled() {
+            return Poll::Ready(take());
+        }
+        Poll::Pending
+    }
+
+    /// Retrieve the exception thrown in the associated coroutine.
+    pub async fn cancelled(&mut self) -> PyObject {
+        poll_fn(|cx| self.poll_cancelled(cx)).await
```

Comment:
Since this probably quite likely to be used in `select!` statements, I wonder whether it is reasonable to use a custom type for the future here which actually implements `FusedFuture` directly (thereby avoiding the need to call `.fuse()` and turning it into a no-op in generic context)?

Would this require more state than just checking the pointer? Since it might just be a single bit, do we do pointer tagging else and could apply that here as well?

Reply:
`FusedFuture` semantic is quite questionable, see https://github.com/rust-lang/futures-rs/issues/2779 (I encountered https://github.com/rust-lang/futures-rs/issues/2455 myself while working on pyo3-async...).
That's why I prefer to forget about `FusedFuture` here. But I'm a tokio user, I never use `FusedFuture` at all, so my opinion may be biased.

Reply:
It is an optimization without requiring API breakage, so we can investigate whenever anyone asks for it if there are an reservations towards the trait. Let's focus on getting the fundamentals right instead then.